### DISCUSSION TITLE
"subprocess-exited-with-error" while installing psycopg_c

### DISCUSSION THREAD

**Discussion by TeleVoyant** [cookiecutter/cookiecutter-django#5325] [Q&A] [answered]
## What happened?
command:
```bash
python -m pip install -r requirements/base.txt && python -m pip install -r requirements/local.txt 
```
does not complete successfully, Error:
![Screenshot from 2024-08-25 18-50-51](https://github.com/user-attachments/assets/ef57bce6-1062-4431-b868-2055c613de39)

## What should've happened instead?
complete modules installation on venv

## Additional details
tried different cookiecutter configurations, ended with the same result (postgres version chosen is 16)

- Host system configuration: Linux hp-zbook-studio 6.6.41 NixOS SMP PREEMPT_DYNAMIC Thu Jul 18 11:21:27 UTC 2024 x86_64 GNU/Linux

  - Version of cookiecutter CLI (get it with `cookiecutter --version`): Cookiecutter 2.6.0 from /home/me/path/to/my/application/venv/lib/python3.11/site-packages (Python 3.11.9 (main, Apr  2 2024, 08:25:04) [GCC 13.2.0])
 
  - OS name and version: NixOS 24.05.3164.63d37ccd2d17 (Uakari) x86_64

    On Linux, run

    ```bash
    lsb_release -a 2> /dev/null || cat /etc/redhat-release 2> /dev/null || cat /etc/*-release 2> /dev/null || cat /etc/issue 2> /dev/null
    ```

    On MacOs, run

    ```bash
    sw_vers
    ```

    On Windows, via CMD, run

    ```
    systeminfo | findstr /B /C:"OS Name" /C:"OS Version"
    ```

    ```bash
    # Insert here the OS name and version
    DISTRIB_CODENAME=uakari
    DISTRIB_DESCRIPTION="NixOS 24.05 (Uakari)"
    DISTRIB_ID=nixos
    DISTRIB_RELEASE="24.05"
    LSB_VERSION="24.05 (Uakari)"
    ANSI_COLOR="1;34"
    BUG_REPORT_URL="https://github.com/NixOS/nixpkgs/issues"
    BUILD_ID="24.05.3164.63d37ccd2d17"

    ```

  - Python version, run `python3 -V`: 3.11.9
  - Docker version (if using Docker), run `docker --version`:
  - docker compose version (if using Docker), run `docker compose --version`:
  - ...

- Options selected and/or [replay file](https://cookiecutter.readthedocs.io/en/latest/advanced/replay.html):
  On Linux and macOS: `cat ${HOME}/.cookiecutter_replay/cookiecutter-django.json`
  (Please, take care to remove sensitive information)

```json
"cookiecutter": {
    "project_name": "LawFirmMSS",
    "project_slug": "lawfirmmss",
    "description": "backend for law firm management system stripped",
    "author_name": "Daniel Tumaini",
    "domain_name": "example.com",
    "email": "mymail@gmail.com",
    "version": "0.1.0",
    "open_source_license": "MIT",
    "username_type": "email",
    "timezone": "+3",
    "windows": "n",
    "editor": "VS Code",
    "use_docker": "n",
    "postgresql_version": "16",
    "cloud_provider": "None",
    "mail_service": "Mailgun",
    "use_async": "n",
    "use_drf": "n",
    "frontend_pipeline": "None",
    "use_celery": "n",
    "use_mailpit": "n",
    "use_sentry": "n",
    "use_whitenoise": "y",
    "use_heroku": "n",
    "ci_tool": "Github",
    "keep_local_envs_in_vcs": "y",
    "debug": "y",
    "_template": "https://github.com/pydanny/cookiecutter-django",
    "_output_dir": "/home/user/path/to/my/project/dir",
    "_repo_dir": "/home/user/.cookiecutters/cookiecutter-django",
    "_checkout": null
  },
```

<summary>
Logs:
<details>
<pre>
$ cookiecutter https://github.com/cookiecutter/cookiecutter-django
project_name [Project Name]: ...
</pre>
</details>
</summary>

**Comment by luzfcb** [comment]
@TeleVoyant, did you ensure that you installed all non-Python dependencies like `libpq-dev` and `build-essential` before running `pip`? 

For Debian, these are the dependencies: https://github.com/cookiecutter/cookiecutter-django/blob/master/%7B%7Bcookiecutter.project_slug%7D%7D/utility/requirements-bookworm.apt

**Comment by TeleVoyant** [comment]
the error occurs as i try to install packages from requirements.txt, can you help me understand how it is related to cookiecutter? (as it completed its prior operation successfully)

**Comment by luzfcb** [comment, marked as answer]
The error occurs because non-Python dependencies needed to install psycopg are missing.

psycopg is a library with several parts written in C language, making it possible to use the libpq library in Python; because of this, to install it using pip, you need the C library compilation toolkit (gcc, make, gnu binutils, and others ) and the source headers of the libpq library.

To solve the problem, simply install the NixOS equivalent packages, which provide the same software and libraries as the Debian [build-essential](https://packages.debian.org/bookworm/build-essential) and [libpq-dev](https://packages.debian.org/bookworm/libpq-dev) packages.
